rust-src/bulletproofs/src/set_membership_proof.rs

```diff
@@ -0,0 +1,960 @@
+    // e_tilde <- a_tilde + s_tilde * x
+    let mut e_tilde = s_tilde;
+    e_tilde.mul_assign(&x);
+    e_tilde.add_assign(&a_tilde);
+    // append tx, tx_tilde, e_tilde to transcript
+    transcript.append_message(b"tx", &tx);
+    transcript.append_message(b"tx_tilde", &tx_tilde);
+    transcript.append_message(b"e_tilde", &e_tilde);
+
+    // Part 5: Inner product proof for tx = <lx,rx>
+    // get challenge w from transcript
+    let w: C::Scalar = transcript.challenge_scalar::<C, _>(b"w");
+    // get generator q
+    let Q = B.mul_by_scalar(&w);
+    // compute scalars c such that c*H = H', that is H_prime_scalars = (1, y^-1,..,
+    // y^-(n-1))
+    let y_inv = match y.inverse() {
+        Some(inv) => inv,
+        None => return Err(ProverError::DivisionError),
+    };
+    let H_prime_scalars = z_vec(y_inv, 0, n);
+    // compute inner product proof
+    let proof =
+        prove_inner_product_with_scalars(transcript, &G, &H, &H_prime_scalars, &Q, &lx, &rx);
+
+    // return range proof
+    if let Some(ip_proof) = proof {
+        return Ok(SetMembershipProof {
+            A,
+            S,
+            T_1,
+            T_2,
+            tx,
+            tx_tilde,
+            e_tilde,
+            ip_proof,
+        });
+    }
+    Err(ProverError::InnerProductProofFailure)
+}
+
+/// Error messages detailing why proof verification failed
+#[derive(Debug, PartialEq, Eq)]
+pub enum VerificationError {
+    /// The set must have a size of a power of two
+    SetSizeNotPowerOfTwo,
+    /// The length of G_H was less than |S|, which is too small
+    NotEnoughGenerators,
+    /// The consistency check for t_0 failed
+    InconsistentT0,
+    /// Choice of randomness led to verification failure
+    DivisionError,
+    /// inner product proof verification failed
+    IPVerificationError,
+}
+
+/// This function verifies a set membership proof, i.e. a proof of knowledge
+/// of value v that is in a set S and that is consistent
+/// with a commitment V to v. The arguments are
+/// - S - the set as a vector
+/// - V - commitment to v
+/// - proof - the set membership proof to verify
+/// - gens - generators containing vectors G and H both of length at least |S|
+///   (bold g,h in bluepaper)
+/// - v_keys - commitment keys B and B_tilde (g,h in bluepaper)
+#[allow(non_snake_case)]
+pub fn verify<C: Curve>(
+    transcript: &mut RandomOracle,
+    the_set: &[u64],
+    V: &Commitment<C>,
+    proof: &SetMembershipProof<C>,
+    gens: &Generators<C>,
+    v_keys: &CommitmentKey<C>,
+) -> Result<(), VerificationError> {
+    // Part 1: Setup
+    let n = the_set.len();
+    if !n.is_power_of_two() {
+        return Err(VerificationError::SetSizeNotPowerOfTwo);
+    }
+    if gens.G_H.len() < n {
+        return Err(VerificationError::NotEnoughGenerators);
+    }
+
+    // TODO: Check whether n fits into u64
+    // TODO: Check whether u64 integers fit into the field (note this should be the
+    // case assuming reasonable fields)
```

Comment:
THere is nothing to check here. That is assumed by C being a Curve, i.e., implementing that trait. Remove this TODO.